lib/srv/db/ca.go

```diff
@@ -172,22 +183,107 @@ func (s *Server) getCACertPaths(database types.Database) ([]string, error) {
+func (s *Server) saveCACert(filePath string, content []byte, version []byte) error {
+	// Save CA contents.
+	err := os.WriteFile(filePath, content, teleport.FileMaskOwnerOnly)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Save the CA version.
+	err = os.WriteFile(filePath+versionFileSuffix, version, teleport.FileMaskOwnerOnly)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.log.Debugf("Saved CA certificate %v.", filePath)
+	return nil
+}
+
+// updateCACert updates the database CA contents if it has changed.
+func (s *Server) updateCACert(ctx context.Context, database types.Database, filePath string) error {
+	var contents []byte
+
+	// Get the current CA version.
+	version, err := s.cfg.CADownloader.GetVersion(ctx, database, filePath)
+	// If getting the CA version is not supported, download it.
+	if trace.IsNotImplemented(err) {
+		contents, version, err = s.cfg.CADownloader.Download(ctx, database, filePath)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	} else if err != nil {
+		return trace.Wrap(err)
+	}
+
+	equal, err := isCAVersionEqual(filePath, version)
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+
+	if equal {
+		return nil
+	}
+
+	// Check if the CA contents were already downloaded. If not, download them.
+	if contents == nil {
+		contents, version, err = s.cfg.CADownloader.Download(ctx, database, filePath)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	err = s.saveCACert(filePath, contents, version)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.log.Infof("Database %q CA updated.", database.GetName())
+	return nil
+}
+
+// isCAVersionEqual compares the in disk version with the provided one.
+func isCAVersionEqual(filePath string, version []byte) (bool, error) {
+	currentVersion, err := os.ReadFile(filePath + versionFileSuffix)
+	if err != nil {
+		return false, trace.ConvertSystemError(err)
+	}
+
+	return bytes.Equal(currentVersion, version), nil
+}
+
 // CADownloader defines interface for cloud databases CA cert downloaders.
 type CADownloader interface {
 	// Download downloads CA certificate for the provided database instance.
-	Download(context.Context, types.Database, string) ([]byte, error)
+	Download(context.Context, types.Database, string) ([]byte, []byte, error)
+	// GetVersion returns the CA version for the provided database.
+	GetVersion(context.Context, types.Database, string) ([]byte, error)
 }
 
 type realDownloader struct {
+	// httpClient is the HTTP client used to download CA certificates.
+	httpClient *http.Client
+	// sqladminClient is the Cloud SQL Admin API service used to download CA
+	// certificates.
+	sqlAdminClient gcp.SQLAdminClient
 }
 
 // NewRealDownloader returns real cloud database CA downloader.
-func NewRealDownloader() CADownloader {
-	return &realDownloader{}
+func NewRealDownloader(ctx context.Context) (CADownloader, error) {
+	sqlAdminClient, err := gcp.NewSQLAdminClient(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```

Comment:
We should only create GCP client when it's needed. This would prevent db service from starting when there is no GCP credentials.

Reply:
Yeah, you're right. I'm going to change that.

Reply:
I've updated to have a similar behavior we have today: Initialize a GCP client for every request. The only change is that it supports using the client on the struct (which will be used only on tests).

Reply:
👍 looks good. Alternatively you can also pass in CloudClients which provides the caching if you want.